source/microbit/modaudio.cpp

```diff
@@ -0,0 +1,430 @@
+        timer->SHORTS = 0;
+    }
+    return audio_reset();
+}
+MP_DEFINE_CONST_FUN_OBJ_0(audio___init___obj, audio_init);
+
+
+STATIC mp_obj_t start() {
+    audio_start();
+    return mp_const_none;
+}
+MP_DEFINE_CONST_FUN_OBJ_0(microbit_audio_start_obj, start);
+
+STATIC mp_obj_t stop() {
+    audio_stop();
+    return mp_const_none;
+}
+MP_DEFINE_CONST_FUN_OBJ_0(microbit_audio_stop_obj, stop);
+
+STATIC mp_obj_t play(mp_uint_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_source, MP_ARG_REQUIRED | MP_ARG_OBJ, {.u_obj = MP_OBJ_NULL} },
+        { MP_QSTR_wait,  MP_ARG_BOOL, {.u_bool = true} },
+        { MP_QSTR_pins,   MP_ARG_OBJ, {.u_obj = mp_const_none } },
+    };
+    // parse args
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_args, pos_args, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+    mp_obj_t src = args[0].u_obj;
+    mp_obj_t pins = args[2].u_obj;
+    if (pins == mp_const_none) {
+        audio_set_pins((mp_obj_t)&microbit_p0_obj, (mp_obj_t)&microbit_p1_obj);
+    } else {
+        mp_obj_t len = mp_obj_len(pins);
+        if (mp_obj_get_int(len) != 2) {
+            nlr_raise(mp_obj_new_exception_msg(&mp_type_ValueError, "wrong number of pins"));
+        }
+        mp_obj_t pin0 = mp_obj_subscr(pins, MP_OBJ_NEW_SMALL_INT(0), MP_OBJ_SENTINEL);
+        mp_obj_t pin1 = mp_obj_subscr(pins, MP_OBJ_NEW_SMALL_INT(1), MP_OBJ_SENTINEL);
+        audio_set_pins(pin0, pin1);
+    }
+    audio_play_source(src, args[1].u_bool);
+    return mp_const_none;
+}
+MP_DEFINE_CONST_FUN_OBJ_KW(microbit_audio_play_obj, 0, play);
+
+microbit_audio_frame_obj_t *new_microbit_audio_frame(void);
+
+STATIC mp_obj_t microbit_audio_frame_new(const mp_obj_type_t *type_in, mp_uint_t n_args, mp_uint_t n_kw, const mp_obj_t *args) {
+    (void)type_in;
+    (void)args;
+    mp_arg_check_num(n_args, n_kw, 0, 0, false);
+    return new_microbit_audio_frame();
+}
+
+STATIC mp_obj_t audio_frame_subscr(mp_obj_t self_in, mp_obj_t index_in, mp_obj_t value_in) {
+    microbit_audio_frame_obj_t *self = (microbit_audio_frame_obj_t *)self_in;
+    mp_int_t index = mp_obj_get_int(index_in);
+    if (index < 0 || index >= AUDIO_CHUNK_SIZE) {
+         nlr_raise(mp_obj_new_exception_msg(&mp_type_ValueError, "index out of bounds"));
+    }
+    if (value_in == MP_OBJ_NULL) {
+        // delete
+        nlr_raise(mp_obj_new_exception_msg(&mp_type_TypeError, "Cannot delete elements of AudioFrame"));
+    } else if (value_in == MP_OBJ_SENTINEL) {
+        // load
+        return MP_OBJ_NEW_SMALL_INT(self->data[index]);
+    } else {
+        mp_int_t value = mp_obj_get_int(value_in);
+        if (value < -128 || value >= 128) {
+            nlr_raise(mp_obj_new_exception_msg(&mp_type_ValueError, "value out of range"));
+        }
+        self->data[index] = value;
+        return mp_const_none;
+    }
+}
+
+const mp_obj_type_t microbit_audio_frame_type = {
```

Comment:
AudioFrame looks exactly like a bytearray (actually array.array('b', ...)).  Why not just use array.array?